packages/next/server/response-cache/web.ts

```diff
@@ -0,0 +1,129 @@
+import type { ResponseCacheEntry, ResponseGenerator } from './types'
+
+/**
+ * In the web server, there is currently no incremental cache provided and we
+ * always SSR the page.
+ */
+export default class WebResponseCache {
+  pendingResponses: Map<string, Promise<ResponseCacheEntry | null>>
+  previousCacheItem?: {
+    key: string
+    entry: ResponseCacheEntry | null
+    expiresAt: number
+  }
+  minimalMode?: boolean
+
+  constructor(minimalMode: boolean) {
+    this.pendingResponses = new Map()
+    this.minimalMode = minimalMode
+  }
+
+  public get(
+    key: string | null,
+    responseGenerator: ResponseGenerator,
+    context: {
+      isManualRevalidate?: boolean
+      isPrefetch?: boolean
+    }
+  ): Promise<ResponseCacheEntry | null> {
+    // ensure manual revalidate doesn't block normal requests
+    const pendingResponseKey = key
+      ? `${key}-${context.isManualRevalidate ? '1' : '0'}`
+      : null
+
+    const pendingResponse = pendingResponseKey
+      ? this.pendingResponses.get(pendingResponseKey)
+      : null
+    if (pendingResponse) {
+      return pendingResponse
+    }
+
+    let resolver: (cacheEntry: ResponseCacheEntry | null) => void = () => {}
+    let rejecter: (error: Error) => void = () => {}
+    const promise: Promise<ResponseCacheEntry | null> = new Promise(
+      (resolve, reject) => {
+        resolver = resolve
+        rejecter = reject
+      }
+    )
+    if (pendingResponseKey) {
+      this.pendingResponses.set(pendingResponseKey, promise)
+    }
+
+    let resolved = false
+    const resolve = (cacheEntry: ResponseCacheEntry | null) => {
+      if (pendingResponseKey) {
+        // Ensure all reads from the cache get the latest value.
+        this.pendingResponses.set(
+          pendingResponseKey,
+          Promise.resolve(cacheEntry)
+        )
+      }
+      if (!resolved) {
+        resolved = true
+        resolver(cacheEntry)
+      }
+    }
+
+    // we keep the previous cache entry around to leverage
+    // when the incremental cache is disabled in minimal mode
+    if (
+      pendingResponseKey &&
+      this.minimalMode &&
+      this.previousCacheItem?.key === pendingResponseKey &&
```

Comment:
If previousCacheItem is always empty in web cache we can remove these related logic?